src/engine/PathSearch.cpp

```diff
@@ -0,0 +1,391 @@
+
+    timer.stop();
+    auto buildingTime = timer.msecs();
+    timer.start();
+
+    std::span<const Id> sources =
+        handleSearchSide(config_.sources_, boundSources_);
+    std::span<const Id> targets =
+        handleSearchSide(config_.targets_, boundTargets_);
+
+    timer.stop();
+    auto sideTime = timer.msecs();
+    timer.start();
+
+    auto paths = allPaths(sources, targets, binSearch);
+
+    timer.stop();
+    auto searchTime = timer.msecs();
+    timer.start();
+
+    CALL_FIXED_SIZE(std::array{getResultWidth()},
+                    &PathSearch::pathsToResultTable, this, idTable, paths);
+
+    timer.stop();
+    auto fillTime = timer.msecs();
+    timer.start();
+
+    auto& info = runtimeInfo();
+    info.addDetail("Time to build graph & mapping", buildingTime.count());
+    info.addDetail("Time to prepare search sides", sideTime.count());
+    info.addDetail("Time to search paths", searchTime.count());
+    info.addDetail("Time to fill result table", fillTime.count());
+  }
+
+  return {std::move(idTable), resultSortedOn(), subRes->getSharedLocalVocab()};
+};
+
+// _____________________________________________________________________________
+VariableToColumnMap PathSearch::computeVariableToColumnMap() const {
+  return variableColumns_;
+};
+
+std::span<const Id> PathSearch::handleSearchSide(
+    const SearchSide& side, const std::optional<TreeAndCol>& binding) const {
+  std::span<const Id> ids;
+  bool isVariable = std::holds_alternative<Variable>(side);
+  if (isVariable && binding.has_value()) {
+    ids = binding->first->getResult()->idTable().getColumn(binding->second);
+  } else if (isVariable) {
+    return {};
+  } else {
+    ids = std::get<std::vector<Id>>(side);
+  }
+  return ids;
+}
+
+// _____________________________________________________________________________
+std::vector<Path> PathSearch::findPaths(
+    const Id source, const std::unordered_set<uint64_t>& targets,
+    const BinSearchWrapper& binSearch) const {
+  std::forward_list<Edge> edgeStack;
+  Path currentPath;
+  std::unordered_map<uint64_t, std::vector<Path>> pathCache;
+  std::unordered_set<uint64_t> visited;
+
+  auto addToCache =
+      [](std::unordered_map<uint64_t, std::vector<Path>>& pathCache,
+         const Path& path, size_t stopIndex) {
+        for (size_t i = 0; i < stopIndex; i++) {
+          const auto& edge = path.edges_[i];
+          auto startIndex = edge.start_.getBits();
+          pathCache.try_emplace(startIndex, std::vector<Path>());
+          pathCache[startIndex].push_back(path.startingAt(i));
+        }
+      };
+
+  visited.insert(source.getBits());
+  for (auto edge : binSearch.outgoingEdes(source)) {
+    edgeStack.push_front(std::move(edge));
+  }
+
+  while (!edgeStack.empty()) {
+    auto edge = edgeStack.front();
+    edgeStack.pop_front();
+
+    while (!currentPath.empty() && edge.start_ != currentPath.end()) {
+      currentPath.pop_back();
+    }
```

Comment:
I think it works that you just do a single pop in the end as soon as you have no more outgoing edges that lead to unvisited nodes. Because then you know, that the next edge will be a